src/main/java/com/southernstorm/noise/protocol/AESGCMFallbackCipherState.java

```diff
@@ -187,7 +187,7 @@ public int encryptWithAd(byte[] ad, byte[] plaintext, int plaintextOffset,
 		int space;
 		if (ciphertextOffset < 0 || ciphertextOffset > ciphertext.length)
 			throw new IllegalArgumentException();
-		if (length < 0 || plaintextOffset < 0 || plaintextOffset > plaintext.length || plaintextOffset + length > plaintext.length)
+		if (length < 0 || plaintextOffset < 0 || plaintextOffset > plaintext.length || plaintextOffset + length < plaintextOffset || plaintextOffset + length > plaintext.length)
```

Comment:
A simpler approach may be to use:
|| length > plaintext.length || (plaintext.length - plaintextOffset) < length
This avoids the wrap-around problem with "plaintextOffset + length" going negative.

Reply:
Nice one, I've implemented your suggestion :)